### PR TITLE
Make demos build-able from top level makefile and add them to CI

### PR DESCRIPTION
**Pull request by nhorman**
We should be ensuring our demos in the source tree are buildable on a regular basis.  Add a target to build them from the top level makefile and add a step to the appropriate CI jobs to check that

### CODE REVIEW

**Review by paulidale — CHANGES_REQUESTED** [review]
Looks good, at least superficially.

**Comment by nhorman** [comment]
@paulidale thanks, not done yet of course, need to fix up all the build errors, and figure out how to link to external libraries using our build system for the http3 demo

**Comment by romen** [comment]
I like what I see, and it makes sense, but I am wondering if you thought of a way to do this without removing the Make files that end users could use in the "here is a simple self-contained demo openssl application that you can try to get started" setting.
I don't have solutions to propose at this time, but I would like a solution that allows both self-contained demo apps and testing them in our integrated build system so they stay current!

**Comment by nhorman** [comment]
@romen this big issue I've run into there is the makefiles themselves.  The makefiles currently use several features that POSIX/GNUmake only.  Windows uses nmake in our build environment, which is, to say the least, limited.  There is no good way I can find to write a Makefile that is compatible with both, that doesn't amount to just writing two parallel makefiles (one for windows, one for Gnu Make), which to me doesn't seem like a much better solution.  

I think, given that we only distribute our demos with our source tree, and the current build (which only works on unix systems at the moment), explicitly link with the libraries that are built from the same source tree, I feel like using the standard build system to enable windows compatibility is a step forward.

I'm open to suggestions of course.  One possibility might be to augment the readme file in each demo directory with instructions on how to build these demos (i.e. what include/library paths are required), to guide users in writing their own independent makefiles specific to their platform.  Would that be sufficient to enable independent builds?

**Comment by slontis** [comment]
Maybe an OTC discussion?
The makefiles arent ideal, but I assume they existed in that form for a reason...
Integrating into the build system, doesnt seem to make it easier for a user to pull out the demo code and try it for themselves.

**Comment by nhorman** [comment]
I can't speak to the history of how/why they were created, but I can say they currently only work on unix systems (i.e. windows users are left writing their own name files for these demos at the moment).  But yeah, we can discuss on OTC

**Comment by nhorman** [comment]
OTC:  what desired build use cases do we want to support for our demo directory?

I.e. is it sufficient to use our perl based build system to build these demos, or is it preferable to use an external makefile constitution?


Advantages to using our build system are that it's already cross platform, and can be easily integrated to our CI

Advantages of an alternative make system (like cmake) are that it would likely be more understandable to users who attempt to build said demos outside of our source tree

**Comment by t8m** [comment]
Could we have both? I.e. keep the existing makefiles in the tree but also include demos in build. Of course that would not test the existing makefiles...

**Comment by nhorman** [comment]
Yeah, I thought about that, but decided against when it became clear that doing so would require updating both the make files and build.info when making some changes.  It also still leaves windows out in the cold due to make incompatiblities.  I think the only non-perl solution here that gives us some level of cross platform support is to drop the existing make files and use cmake or meson/ninja to build these targets

**Comment by t-j-h** [comment]
There is no requirement for the demos to be separately buildable. We just didn't previously maintain them so we never plugged them into the build setup. They should be plugged in - using the same build approach we use for everything else.

This also isn't something we need to ask the OTC about ...

**Comment by tom-cosgrove-arm** [comment]
> Advantages of an alternative make system (like cmake) are that it would likely be more understandable to users who attempt to build said demos outside of our source tree

I don't see the benefit in supporting this. Someone building OpenSSL itself has to use our build system; it seems reasonale to require this for the demos - I agree with @t-j-h

**Comment by paulidale** [comment]
I think they should be buildable via our build system but using the Makefiles contained in the demos directory rather than creating something special ourselves.

**Comment by nhorman** [comment]
I think its critical to recognize two things here:
1) The makefiles as they exist currently, don't allow for building outside the current source tree ( they are written with C/LDFLAGS hard coded to assume that libcrypo/libssl/header files are located in the same root source tree as the demos
2) the makefiles as they exist currently only support unix and unix-like targets - windows cannot use them in any way as nmake can't reliably be used with them in a compatible manner

As such, we should recognize that any solution that (a) allows for building outside the source tree and (b) allows for cross platform building will require a complete rewrite of these makefiles, and likely require conversion to another build system entirely (cmake or meson/ninja) to support non-unix platforms in a single build directive file.

Given the above, we should also recognize that anyone who has used these demos up to now outside of our source tree has done so by modifying the makefiles to suite their needs.  unix users have had a slight advantage because the makefiles could just be modified slightly to get something building, but windows users have necessarily been doing a complete rewrite to be nmake compatible.

As such, while I'm not opposed to making these buildable from outside the source tree, I'm struggling to see  a community driven need to do so.  Anyone who has used them up to now would be in no worse a position than they were previously by switching to our perl-based build system, but for those who are building within the soruce tree, they would be better off because, by switching to our internal build system, we now become cross platform enabled (which also enables our CI to build the demos, being the initial purpose of this PR).

thinking about it another way for comparison, our unit test infrastructure is occasionally used outside of our source tree (in particular I believe evp_extra_test is often used for various purposes), but we don't provide external makefiles for those.  I'm not sure I see a strong need from the community to treat the demos directory any differently

**Comment by romen** [comment]
@nhorman my point of view is that, with the current `Makefile` approach, I can instruct a B.Sc. student to bootstrap a simple application by taking one of the demo apps, spending a few hours on StackOverflow/Reddit to try and figure out how to "fix" the Makefile for their needs, and have a reasonable chance of them getting to a good starting point for a new toy project relatively quickly.

Granted, we have it easy because they are usually using Linux for this kind of activity.

The way I see it, **removing** the Makefiles in favor of our perl-based build-system is going to create a new high barrier to entry for this kind of use of the programs in the demo tree, and it was my understanding that we wanted to support this kind of use for newcomers: our custom build system is often quite exotic also for experienced OpenSSL veterans, and Perl is one more thing that newcomers would need to learn, before they could use these demos in their own experiments.

I do also understand that the demos have little value if we don't regularly test them to ensure they are correctly aligned to the development process, which to me justifies the addition of our build system, I'd just prefer not to lose the current Makefiles.

I do also value your point about duplicated efforts to make sure the Makefiles are in sync with changes in our `build.info`, which makes sense and is indeed a valid argument against keeping the Makefiles in.

I have 2 counterproposals, with different levels of complexity and required work:
- easy, let's keep the existing Makefile there, maybe adding a disclaimer that
  - the Makefile is just an example of how a simple Makefile to compile a demo standalone app could look like in a Linux development environment, but will most likely require tweaks
  - the Makefile itself is untested, as our testing infrastructure uses the `build.info` system for portability reason
- harder, we could generate `Makefile.unix`, `Makefile.nmake`, etc. out of the maintained `build.info` and commit them to the source tree every time there are changes.
  - we could use CI/CD to make sure that the few Makefiles we pre-generate and include in the source tree do not get out of sync by testing that no changes are detected, similarly to what we do for the `make update` step.
  
Anyway, I will try to raise this during the next OTC to see what the consensus is.

Whatever we end up choosing, thanks for the valuable work you did in this PR and everyone for your input in the discussion!

**Comment by paulidale** [comment]
I don't see any problem with our `Makefile.in` (for suitable targets) including essentially:

```make
demos:
	make -C demos
```

Easy to hook into our build framework to verify that the demos build correctly and it provides a reasonably easy _demo_ build system.  There would have to be more to run the demos & this seems sensible as part of our unit testing & done via Perl.  The various `Makefile` under `demos` should be kept simple & treated as part of the demo code.

**Comment by nhorman** [comment]
@paulidale please see my previous comment.  Hooking into the demo make files isn't hard.  The hard part is for each individual makefile, setting an appropriate include/library path for the tree local libcrypro/SSL by default, but allowing it to be.overridden easily for use.outside the source tree, and doing so in a way that is compatible with both gnumake and nmake.

**Comment by paulidale** [comment]
Could we pass `CFLAGS` et al from our build system?

For out of tree builds, mightn't we just copy the demos sources to the build tree directory?

**Comment by nhorman** [comment]
We could copy the flags sure. But I don't see how that will help matters.  The include paths passed in the top of the source tree when building openssl won't be correct relative to the demos/{demo} subdirectories.

As for copying the source....I'm a bit confused about the purpose.  The demos are already in the openssl tree. If the desire was to be able.to build them outside the source tree, it seems somewhat counterproductive to just have to copy them back into the tree.  I'm sure I'm missing your point here, but this doesn't currently make sense to me.

**Comment by paulidale** [comment]
The point was to copy them to the build tree and confirm that they build there.  There is no reason to copy them back afterwards.

IMO we want to build the demoes as part of our testing but we don't want to complicate them in any way for folks who need the demos.  I.e. the makefiles under demos should not be contaminated and ideally would be further simplified.  Our build/test infrastructure is what should gain any complexity required.  Thus, copying the demos to the build tree and building them with an appropriate environment set seems very much like the way to go to verify they build (& run?) okay.

I'm expecting users who need the demos are likely to be doing in tree builds while they're learning.  Once they are confident in building, they ought to be able to incorporate demo code into their apps.  I.e. users being able to build specific demos from their build tree is not a requirement IMO.

So a test recipe that stuffs the environment, maybe copies the sources and finally runs `make -C` is good enough for our CI.  What we've got currently in demos is good enough as a demo.  Each executable is pretty straightforward to build and the information content is in the .c files not the Makefile.

**Comment by nhorman** [comment]
@romen can I ask you to clarify something please?  You indicated that you have students that you instruct to use the openssl demos to start a project, which they as a first step spend time 'fixing' the makefile.  Do you consider that 'fixing' step to be part of your students learning experience?  I ask because I don't think any solution we go with here is going to preserve that aspect of the learning experience.  I.e. when this work is done, the build of the demos will 'just work' ideally.

If however the 'fixing' of the makefile is just a necessary evil to get to building the demos in the pursuit of a project, I can certainly understand the desire to have the build directions be intuitive and understandable to a student, so they can expand and alter them.

If we are in agreement on the above point, we may be able to come to a compromise here.  I'm strongly against any multiple build system solution, because I believe practically speaking, we will only use one and the other(s) will suffer bit rot, and I don't think that is productive.

That said, if there is a relatively simple way to translate build.info files to other formats for the purpose of education I'm amenable to that.  I'm not enthusiastic about committing those translated files for the above mentioned but rot concerns, but I think having them generatable I would hope would be sufficient.

I'd also wonder if just updating the readme to verbally note general build instructions wouldn't be sufficient (perhaps even superior to a semi broken makefile).  That is to say, currently every demo we have is a single c file that requires an include path and a library path.  Would it be sufficient to your students if we augmented the readme to verbally explain how to build a given demo in high level terms?

**Comment by slontis** [comment]
When I last updated the demo makefiles, the target goal was to make it so they are buildable, to demonstrate the demos actually build.
Are the makefiles easier to understand for a new user than our build system - I would say yes. (This could be countered by supplying build instructions)

**Comment by romen** [comment]
@nhorman the "fixing" the `Makefile` part is a necessary evil, as I often ask them to take the source from the OpenSSL tree and move it to a new git repo from scratch to build against the system installed OpenSSL.

I expect that, after this treatment, once we reach a "it just works" in-tree build, they will likely still need to get their hands dirty to make sure they can build it out-of-tree. It is a valuable learning experience for them to understand differences between assembler, compiler, linker, and the steps of the build process, and how they can alter the build process to tweak those knobs.

Another thing I might ask them once they are done with this first step, is to then build (or run) the same project against a custom development install of OpenSSL. That way they get to experience the runtime dynamic loader tweaks, `LD_LIBRARY_PATH`, rpath/runpath, etc.; they also see the difference between having debugging symbols and sources available only for their toy project and when they have debug symbols and sources for OpenSSL as well.

I am not asking OpenSSL to support my very niche use case, and it's fine by me if they have to edit the Makefile to get things working out-of-tree. As @slontis mentioned, for students or new users discovering how to tweak a Makefile for these needs is relatively trivial.
If they have to deal with `build.info` they are limited only to the scarce documentation we provide, little to no "tutorials"/"questions" resources in the wider Internet for common tasks, and they would need to understand Perl to investigate how the `build.info` entries get translated into a `Makefile`.

Most likely if we lose the simple `Makefile`s as part of this PR, in the future I might just have to ask them to create a Makefile from scratch rather than tweaking an existing one. So, it's not the end of the world if we lose them, but I can see as my specific use case might be of value for other scenarios where we want these demos to be easy to access for newcomers that want to rapidly prototype a simple toy project built on top of OpenSSL.

**Comment by paulidale** [comment]
@romen, I don't think you use case is niche.  Rather, I think it's the one of the two common cases for demo.  The other is build and running the demo directly just to see what it does & how it works.  This case transitions naturally into wrapping other code around the demo core (which is essentially your case).

**Comment by nhorman** [comment]
@romen @paulidale I'm sympathetic to the use cases described, but I'm having a hard time aligning the needs of that use case in conjunction with the needs to ensure that we have demos that are build-able in a cross platform fashion in our CI environment. 

The top level problem as I see it is that Makefiles aren't writeable in a manner that is reliably usable in both a unix and windows environment (which is needed for CI testing)

We could convert to using meson or cmake, which would give us cross platform usability, but would require that individuals install additional tools when building these within the source tree, which they may not want to do. I'm not super comfortable with the idea of creating new build requirements just for these demos.  It would additionally require that our CI infrastructure be updated to install the requisite tooling to build them.

In pursuit of the goal in this related issue, I really think using our available build system is the right approach.  But if the common use case is to use them demos independently of the openssl source tree, then perhaps the longer term solution is to _not_ include the demos in the source tree at all, instead maintaining them in their own repository in the openssl project with their own build system and CI to ensure they work consistently.  Thats a larger undertaking however, and bears some further discussion around what our policies might be for this code.

**Comment by nhorman** [comment]
@beldmit That actually pretty well covers everything we need (all our platforms support gnu/posix makefiles save for windows that supports nmake, which is simmilar but not in any meaningful way compatible).  So to support both of those platforms, what we need is either:

1) a build system that supports both
or 
2) separate makefiles (one for unix, one for windows)
or
3) Another alternate build system that supports all platforms

(1) is available, and is our current perl based build system
(2) isn't really sensible, as it means we need to maintain 2 separate makefile system, separate and distinct from our own perl based build system (which seems like a recipe for failing to update one or more of them) 
(3) could be done (i.e. cmake or meson, as those can target windows), but requires that we install those tools as part of our build requirements just to build the demos.  

The only drawback that I see to using our in-place build system is the fact that we have this use case, in which students (or anyone who does not have familiarity with our build system) might find altering it to extend the demos daunting.

Perhaps the answer then is "Make our build system be less daunting".  theres not a need to understand the full build system, I think, just to understand how to interact with it through the build.info files.  Perhaps a top level readme here would help that explains the directives in a build.info file, and what you need to do to extend them?

**Comment by tom-cosgrove-arm** [comment]
It seems that getting the demos to build using our current multi-platform system would be the most logical thing to do. It meets our immediate need to build in the CI, and allows other people to build then by downloading the source and configuring and running `make`. Requiring an additional `Makefile` generator like `cmake` would seem to be an anti-pattern at this point.

There are pros (a starting point for some people) and cons (they are likely to bit-rot) to keeping any existing `Makefile`s around. Maybe if we renamed to something like `Makefile.example.unix` it would at least give the hint that they're not first-class citizens

**Comment by paulidale** [comment]
Our existing build infrastructure will be independent of the demo Makefiles.  We create one monster Makefile.  So the two aren't necessarily incompatible.  Adding build.info files to demos might be enough to test that the demos compile.  What it won't test is that the Makefiles will build.  It wouldn't be ideal but perhaps we can live with this?

Another approach is for a test to do the build using the Makefiles, it would just be one extra step in the workflow.  No need to worry about cross platform or different OSes.

Perhaps do both: i.e. add build.info files and a test that builds via make.  The build.info would have a comment to ignore them, they are for internal testing purposes (or better have a build.info that isn't in the demo directories if it possible).  The test would be in tree and on Linux and might even be able to run some of the outputs.

**Comment by t8m** [comment]
The test that would test the demo makefiles could be just something that runs in a Github workflow file. We do not even need to pollute the top level makefile with such target.

So yeah, having both: build.info in demos to test cross-platform build and a separate make call in some CI job that tests the Makefiles in demos would be fine. We can add a comment into these makefiles that the expected build environment for these makefiles is Linux and other platform will require tweaks to make them usable.

File: demos/bio/build.info [hunk 1]
```diff
@@ -0,0 +1,29 @@
+PROGRAMS{noinst}=client-arg sconnect server-cmod client-conf saccept server-arg server-conf
```
Comment (levitte) [inline demos/bio/build.info:1]: You did remember it in a few places, like here here, but......

File: demos/cms/build.info [hunk 2]
```diff
@@ -0,0 +1,45 @@
+PROGRAMS = cms_comp \
```
Comment (levitte) [inline demos/cms/build.info:1]: .... not here

**Review by levitte — CHANGES_REQUESTED** [review]
Unless you want a surprise when installing, you should change `PROGRAM =` to `PROGRAM{noinst} =` everywhere.

**Comment by levitte** [comment]
I'm a bit apprehensive of this PR, because the demos Makefiles are also a demonstration on how to build simple apps on your own.  When moving all the building info to `build.info`, the demos become more integral to building OpenSSL, and thereby lose that aspect of the demonstration.

**Comment by tom-cosgrove-arm** [comment]
So it appears the consensus is to have both: incorporate the demos into the main build system (but not install them!) so we can check that the source code for the demos doesn't bit rot.

But also keep the demo `Makefile`s. We can then (separately to this PR?) work on a CI job that checks that they work in a Unix-type environment. And accept contributions for `Makefile.VS`s (or equivalent) for Windows

**Comment by levitte** [comment]
I can deal with that

File: demos/bio/build.info (hunk 1 above)
Comment (nhorman) [reply demos/bio/build.info:1]: I've reconciled all of these to be {noinst}

**Comment by nhorman** [comment]
Ok, still working on getting the windows builds solid, but I've restored all the makefiles.  They are currently unused in our build system (we use the perl build system for that), but they exist as examples for guidance

**Comment by tom-cosgrove-arm** [comment]
> I've restored all the makefiles

Thanks.  We should add something in `demos/README.txt` (and should that become a `.md` like all the others?) saying how to build them (the comment in `Configure` is useful, but not where people are likely to look) and also setting expectations about the `Makefile`s (maybe along the lines of "Makefiles are provided for Unix-like systems, but are only infrequently checked")

**Comment by paulidale** [comment]
I like the final suggestion :+1:

**Review by t-j-h — APPROVED** [review]

**Review by romen — APPROVED** [review]
LVGTM! Thanks @nhorman for the work, and everyone for the discussion!

**Review by paulidale — APPROVED** [review]
Great outcome.

**Review by tom-cosgrove-arm — APPROVED** [review]
Thanks @nhorman - nice work

**Comment by nhorman** [comment]
merged

**Pull request closed without merging**